src/authorized_principals/authorized_principals.cc

```diff
@@ -0,0 +1,94 @@
+// Copyright 2023 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <filesystem>
+#include <iostream>
+
+#include <signal.h>
+
+#include <oslogin_utils.h>
+#include <oslogin_sshca.h>
+
+using std::cout;
+using std::endl;
+using std::filesystem::path;
+
+using oslogin_sshca::FingerPrintFromBlob;
+
+using oslogin_utils::AuthOptions;
+using oslogin_utils::AuthorizeUser;
+using oslogin_utils::CloseSysLog;
+using oslogin_utils::SetupSysLog;
+using oslogin_utils::SysLogErr;
+
+#define SYSLOG_IDENT "sshd"
+#define SUCCESS 0
+#define FAIL    1
+
+void signal_handler(int signo) {
+  _Exit(0);
+}
+
+int main(int argc, char* argv[]) {
+  size_t fp_len;
+  char *user_name, *cert, *fingerprint;
+  struct sigaction sig;
+  struct AuthOptions opts;
+  string user_response, progname = path(string(argv[0])).filename();
+
+  fp_len = 0;
+  opts = { 0 };
+  user_name = cert = fingerprint = NULL;
+
+  SetupSysLog(SYSLOG_IDENT, progname.c_str());
+
+  if (argc != 3) {
+    SysLogErr("usage: %s [username] [base64-encoded cert]", progname.c_str());
+    goto fail;
+  }
+
+  sig = { 0 };
+  sig.sa_handler = signal_handler;
+  sigemptyset(&sig.sa_mask);
+
+  if (sigaction(SIGPIPE, &sig, NULL) == -1) {
+    SysLogErr("Unable to initialize signal handler. Exiting.");
+    goto fail;
+  }
+
+  user_name = argv[1];
+  cert = argv[2];
+
+  fp_len = FingerPrintFromBlob(cert, &fingerprint);
+  if (fp_len == 0) {
+    SysLogErr("Could not extract/parse fingerprint from certificate.");
+    goto fail;
```

Comment:
I think you may want to free(fingerprint) here too, though I suppose the program will terminate soon and get cleaned up anyway, so a memory leak doesn't matter too much here.

Reply:
for sure, we should have a `free()` in the fail block. I'll get that updated. Thanks.

Reply:
oops, I just re-reviewed the code, it actually shouldn't `free(fingerprint);` in the fail block, all the error conditions for `fail:` happen either before having a `fingerprint` allocated or by failing to allocate it.